### PR TITLE
Wait for context init before using eventLoop in WebSocketClient

### LINKED ISSUE #6839: WebSocketClient.connect() throws IllegalStateException: Should call init(endpoint) with an async EndpointGroup
## Description
I think I hit a bug in the WebSocket client. When a WebSocketClient is built over an EndpointGroup whose selectNow() returns null and resolves asynchronously via select() (service-discovery-backed groups, a cold DynamicEndpointGroup, etc.), connect() fails immediately with:
```
java.lang.IllegalStateException: Should call init(endpoint) before invoking this method.
at com.linecorp.armeria.internal.client.DefaultClientRequestContext.eventLoop(DefaultClientRequestContext.java:825)
at com.linecorp.armeria.client.websocket.DefaultWebSocketClient.connect(DefaultWebSocketClient.java:110)
```

A WebClient over the same group works fine (it waits for selection), so it looks specific to the ws path. 
iiuc the cause is the eager use of ctx.eventLoop [here](https://github.com/line/armeria/blob/f25313347428c31ba96ec3793640048901d525c6/core/src/main/java/com/linecorp/armeria/client/websocket/DefaultWebSocketClient.java#L109)
Minimal [repro](https://gist.github.com/aakugan/2ba06ad01ca6ea8b817e1ec72b2f4b53) (jbang, Armeria 1.39.0).

Please correct me if I'm misreading this and async EndpointGroups aren't intended to be supported by the WebSocket client.
I'd like to pick this up and put together a fix, happy to align on the approach first.

### PR DESCRIPTION
**Pull request by kalayciburak**
Motivation:

`WebSocketClient.connect()` calls `ctx.eventLoop()` immediately after `execute()`. When the client is backed by an `EndpointGroup` whose `selectNow()` returns `null` (service-discovery-backed groups, a cold `DynamicEndpointGroup`, etc.), the context is not initialized yet and `connect()` throws `IllegalStateException: Should call init(endpoint) before invoking this method`. A `WebClient` over the same group waits for selection.

Modifications:

- Finish the WebSocket handshake after `ClientRequestContextExtension.whenInitialized()`, as suggested on the issue.
- Throw if the client context is not a `ClientRequestContextExtension`.
- Add a regression test that connects through an initially empty `DynamicEndpointGroup` and checks a successful handshake status (101 or 200).

Result:

- Closes #6839
- `WebSocketClient.connect()` waits for async endpoint selection instead of failing immediately.

Tested:

- `./gradlew :core:test --tests com.linecorp.armeria.client.websocket.WebSocketClientTest`
- `./gradlew :core:checkstyleMain :core:checkstyleTest`

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/client/websocket/WebSocketClientTest.java [hunk 1]
```diff
@@ -116,6 +119,25 @@ void webSocketClient(SessionProtocol protocol, boolean defaultClient) throws Int
...
 
+    @Test
+    void connectWaitsForAsyncEndpointGroup() {
+        final DelayedEndpointGroup group = new DelayedEndpointGroup();
+        final WebSocketClient client = WebSocketClient.builder(SessionProtocol.HTTP, group).build();
+        final CompletableFuture<WebSocketSession> future = client.connect("/chat");
+        assertThat(future).isNotDone();
+        group.set(server.httpEndpoint());
+        final WebSocketSession session = future.join();
+        assertThat(session.responseHeaders().status().code()).isGreaterThanOrEqualTo(100);
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/websocket/WebSocketClientTest.java]: 4xx or 5xx are not something we want. 😉

File: core/src/main/java/com/linecorp/armeria/client/websocket/DefaultWebSocketClient.java [hunk 2]
```diff
@@ -106,6 +107,36 @@ public CompletableFuture<WebSocketSession> connect(String path, HttpHeaders head
...
+            finishConnect(ctx, requestHeaders, response, outboundFuture, result);
+            return result;
+        }
+        ctxExt.whenInitialized().handle((unused, cause) -> {
+            if (cause != null) {
+                response.abort(cause);
+                outboundFuture.completeExceptionally(cause);
+                result.completeExceptionally(cause);
+                return null;
+            }
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/websocket/DefaultWebSocketClient.java]: Probably we don't need these because `whenInitialized()` never completes with an exception.

File: core/src/main/java/com/linecorp/armeria/client/websocket/DefaultWebSocketClient.java [hunk 3]
```diff
@@ -106,6 +107,36 @@ public CompletableFuture<WebSocketSession> connect(String path, HttpHeaders head
             response = webClient.execute(request, requestOptions);
             ctx = captor.get();
         }
+
+        final CompletableFuture<WebSocketSession> result = new CompletableFuture<>();
+        final ClientRequestContextExtension ctxExt = ctx.as(ClientRequestContextExtension.class);
+        if (ctxExt == null) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/websocket/DefaultWebSocketClient.java:112]: I think we can just throw an exception if `ctxExt` is null in this case.

**Review by minwoox — COMMENTED** [review]
Left a few comments. Thanks!

File: core/src/test/java/com/linecorp/armeria/client/websocket/WebSocketClientTest.java [hunk 4]
```diff
@@ -116,6 +120,25 @@ void webSocketClient(SessionProtocol protocol, boolean defaultClient) throws Int
...
 
+    @Test
+    void connectWaitsForAsyncEndpointGroup() {
+        final DelayedEndpointGroup group = new DelayedEndpointGroup();
+        final WebSocketClient client = WebSocketClient.builder(SessionProtocol.HTTP, group).build();
+        final CompletableFuture<WebSocketSession> future = client.connect("/chat");
+        assertThat(future).isNotDone();
+        group.set(server.httpEndpoint());
+        final WebSocketSession session = future.join();
+        assertThat(session.responseHeaders().status()).isIn(HttpStatus.SWITCHING_PROTOCOLS, HttpStatus.OK);
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/websocket/WebSocketClientTest.java]: How about using `@ParameterizedTest`?
```
@ParameterizedTest
@EnumSource(value = SessionProtocol.class, names = {"H2C", "H1C"})
void connectWaitsForAsyncEndpointGroup(SessionProtocol sessionProtocol) {
...
if (sessionProtocol == H1C) {
  assertThat(session.responseHeaders().status()).isSameAs(HttpStatus.SWITCHING_PROTOCOLS);
} else {
  ...
}
```

**Review by minwoox — APPROVED** [review]
Just one nit. 😉

**Review by jrhee17 — APPROVED** [review]
👍 👍

File: core/src/test/java/com/linecorp/armeria/client/websocket/WebSocketClientTest.java (hunk 4 above)
Comment (kalayciburak) [reply core/src/test/java/com/linecorp/armeria/client/websocket/WebSocketClientTest.java]: done, split H1C/H2C statuses
